<p>::::::::::::::::::::::::::::::::::::::: objectives</p>

<ul>
  <li>To be able to read a traceback, and determine where the error took place and what type it is.</li>
  <li>To be able to describe the types of situations in which syntax errors, indentation errors, name errors, index errors, and missing file errors occur.</li>
</ul>

<p>::::::::::::::::::::::::::::::::::::::::::::::::::</p>

<p>:::::::::::::::::::::::::::::::::::::::: questions</p>

<ul>
  <li>How does Python report errors?</li>
  <li>How can I handle errors in Python programs?</li>
</ul>

<p>::::::::::::::::::::::::::::::::::::::::::::::::::</p>

<p>Every programmer encounters errors,
both those who are just beginning,
and those who have been programming for years.
Encountering errors and exceptions can be very frustrating at times,
and can make coding feel like a hopeless endeavour.
However,
understanding what the different types of errors are
and when you are likely to encounter them can help a lot.
Once you know <em>why</em> you get certain types of errors,
they become much easier to fix.</p>

<p>Errors in Python have a very specific form,
called a <a href="../learners/reference.md#traceback">traceback</a>.
Let’s examine one:</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># This code has an intentional error. You can type it directly or
# use it for reference to understand the error message below.
</span><span class="k">def</span> <span class="nf">favorite_ice_cream</span><span class="p">():</span>
    <span class="n">ice_creams</span> <span class="o">=</span> <span class="p">[</span>
        <span class="s">'chocolate'</span><span class="p">,</span>
        <span class="s">'vanilla'</span><span class="p">,</span>
        <span class="s">'strawberry'</span>
    <span class="p">]</span>
    <span class="k">print</span><span class="p">(</span><span class="n">ice_creams</span><span class="p">[</span><span class="mi">3</span><span class="p">])</span>

<span class="n">favorite_ice_cream</span><span class="p">()</span>
</code></pre></div></div>

<pre><code class="language-error">---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
&lt;ipython-input-1-70bd89baa4df&gt; in &lt;module&gt;()
      9     print(ice_creams[3])
      10
----&gt; 11 favorite_ice_cream()

&lt;ipython-input-1-70bd89baa4df&gt; in favorite_ice_cream()
      7         'strawberry'
      8     ]
----&gt; 9     print(ice_creams[3])
      10
      11 favorite_ice_cream()

IndexError: list index out of range
</code></pre>

<p>This particular traceback has two levels.
You can determine the number of levels by looking for the number of arrows on the left hand side.
In this case:</p>

<ol>
  <li>
    <p>The first shows code from the cell above,
  with an arrow pointing to Line 11 (which is <code class="language-plaintext highlighter-rouge">favorite_ice_cream()</code>).</p>
  </li>
  <li>
    <p>The second shows some code in the function <code class="language-plaintext highlighter-rouge">favorite_ice_cream</code>,
  with an arrow pointing to Line 9 (which is <code class="language-plaintext highlighter-rouge">print(ice_creams[3])</code>).</p>
  </li>
</ol>

<p>The last level is the actual place where the error occurred.
The other level(s) show what function the program executed to get to the next level down.
So, in this case, the program first performed a
<a href="../learners/reference.md#function-call">function call</a> to the function <code class="language-plaintext highlighter-rouge">favorite_ice_cream</code>.
Inside this function,
the program encountered an error on Line 6, when it tried to run the code <code class="language-plaintext highlighter-rouge">print(ice_creams[3])</code>.</p>

<p>:::::::::::::::::::::::::::::::::::::::::  callout</p>

<h2 id="long-tracebacks">Long Tracebacks</h2>

<p>Sometimes, you might see a traceback that is very long
-- sometimes they might even be 20 levels deep!
This can make it seem like something horrible happened,
but the length of the error message does not reflect severity, rather,
it indicates that your program called many functions before it encountered the error.
Most of the time, the actual place where the error occurred is at the bottom-most level,
so you can skip down the traceback to the bottom.</p>

<p>::::::::::::::::::::::::::::::::::::::::::::::::::</p>

<p>So what error did the program actually encounter?
In the last line of the traceback,
Python helpfully tells us the category or type of error (in this case, it is an <code class="language-plaintext highlighter-rouge">IndexError</code>)
and a more detailed error message (in this case, it says “list index out of range”).</p>

<p>If you encounter an error and don’t know what it means,
it is still important to read the traceback closely.
That way,
if you fix the error,
but encounter a new one,
you can tell that the error changed.
Additionally,
sometimes knowing <em>where</em> the error occurred is enough to fix it,
even if you don’t entirely understand the message.</p>

<p>If you do encounter an error you don’t recognize,
try looking at the
<a href="https://docs.python.org/3/library/exceptions.html">official documentation on errors</a>.
However,
note that you may not always be able to find the error there,
as it is possible to create custom errors.
In that case,
hopefully the custom error message is informative enough to help you figure out what went wrong.</p>

<p>:::::::::::::::::::::::::::::::::::::::  challenge</p>

<h2 id="reading-error-messages">Reading Error Messages</h2>

<p>Read the Python code and the resulting traceback below, and answer the following questions:</p>

<ol>
  <li>How many levels does the traceback have?</li>
  <li>What is the function name where the error occurred?</li>
  <li>On which line number in this function did the error occur?</li>
  <li>What is the type of error?</li>
  <li>What is the error message?</li>
</ol>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># This code has an intentional error. Do not type it directly;
# use it for reference to understand the error message below.
</span><span class="k">def</span> <span class="nf">print_message</span><span class="p">(</span><span class="n">day</span><span class="p">):</span>
    <span class="n">messages</span> <span class="o">=</span> <span class="p">[</span>
        <span class="s">'Hello, world!'</span><span class="p">,</span>
        <span class="s">'Today is Tuesday!'</span><span class="p">,</span>
        <span class="s">'It is the middle of the week.'</span><span class="p">,</span>
        <span class="s">'Today is Donnerstag in German!'</span><span class="p">,</span>
        <span class="s">'Last day of the week!'</span><span class="p">,</span>
        <span class="s">'Hooray for the weekend!'</span><span class="p">,</span>
        <span class="s">'Aw, the weekend is almost over.'</span>
    <span class="p">]</span>
    <span class="k">print</span><span class="p">(</span><span class="n">messages</span><span class="p">[</span><span class="n">day</span><span class="p">])</span>

<span class="k">def</span> <span class="nf">print_sunday_message</span><span class="p">():</span>
    <span class="n">print_message</span><span class="p">(</span><span class="mi">7</span><span class="p">)</span>

<span class="n">print_sunday_message</span><span class="p">()</span>
</code></pre></div></div>

<pre><code class="language-error">---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
&lt;ipython-input-7-3ad455d81842&gt; in &lt;module&gt;
     16     print_message(7)
     17 
---&gt; 18 print_sunday_message()
     19 

&lt;ipython-input-7-3ad455d81842&gt; in print_sunday_message()
     14 
     15 def print_sunday_message():
---&gt; 16     print_message(7)
     17 
     18 print_sunday_message()

&lt;ipython-input-7-3ad455d81842&gt; in print_message(day)
     11         'Aw, the weekend is almost over.'
     12     ]
---&gt; 13     print(messages[day])
     14 
     15 def print_sunday_message():

IndexError: list index out of range
</code></pre>

<p>:::::::::::::::  solution</p>

<h2 id="solution">Solution</h2>

<ol>
  <li>3 levels</li>
  <li><code class="language-plaintext highlighter-rouge">print_message</code></li>
  <li>13</li>
  <li><code class="language-plaintext highlighter-rouge">IndexError</code></li>
  <li><code class="language-plaintext highlighter-rouge">list index out of range</code> You can then infer that
  <code class="language-plaintext highlighter-rouge">7</code> is not the right index to use with <code class="language-plaintext highlighter-rouge">messages</code>.</li>
</ol>

<p>:::::::::::::::::::::::::</p>

<p>::::::::::::::::::::::::::::::::::::::::::::::::::</p>

<p>:::::::::::::::::::::::::::::::::::::::::  callout</p>

<h2 id="better-errors-on-newer-pythons">Better errors on newer Pythons</h2>

<p>Newer versions of Python have improved error printouts.  If you are debugging errors, it is often
helpful to use the latest Python version, even if you support older versions of Python.</p>

<p>::::::::::::::::::::::::::::::::::::::::::::::::::</p>

<h2 id="syntax-errors">Syntax Errors</h2>

<p>When you forget a colon at the end of a line,
accidentally add one space too many when indenting under an <code class="language-plaintext highlighter-rouge">if</code> statement,
or forget a parenthesis,
you will encounter a <a href="../learners/reference.md#syntax-error">syntax error</a>.
This means that Python couldn’t figure out how to read your program.
This is similar to forgetting punctuation in English:
for example,
this text is difficult to read there is no punctuation there is also no capitalization
why is this hard because you have to figure out where each sentence ends
you also have to figure out where each sentence begins
to some extent it might be ambiguous if there should be a sentence break or not</p>

<p>People can typically figure out what is meant by text with no punctuation,
but people are much smarter than computers.
If Python doesn’t know how to read the program,
it will give up and inform you with an error.
For example:</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="k">def</span> <span class="nf">some_function</span><span class="p">()</span>
    <span class="n">msg</span> <span class="o">=</span> <span class="s">'hello, world!'</span>
    <span class="k">print</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>
     <span class="k">return</span> <span class="n">msg</span>
</code></pre></div></div>

<pre><code class="language-error">  File "&lt;ipython-input-3-6bb841ea1423&gt;", line 1
    def some_function()
                       ^
SyntaxError: invalid syntax
</code></pre>

<p>Here, Python tells us that there is a <code class="language-plaintext highlighter-rouge">SyntaxError</code> on line 1,
and even puts a little arrow in the place where there is an issue.
In this case the problem is that the function definition is missing a colon at the end.</p>

<p>Actually, the function above has <em>two</em> issues with syntax.
If we fix the problem with the colon,
we see that there is <em>also</em> an <code class="language-plaintext highlighter-rouge">IndentationError</code>,
which means that the lines in the function definition do not all have the same indentation:</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="k">def</span> <span class="nf">some_function</span><span class="p">():</span>
    <span class="n">msg</span> <span class="o">=</span> <span class="s">'hello, world!'</span>
    <span class="k">print</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>
     <span class="k">return</span> <span class="n">msg</span>
</code></pre></div></div>

<pre><code class="language-error">  File "&lt;ipython-input-4-ae290e7659cb&gt;", line 4
    return msg
    ^
IndentationError: unexpected indent
</code></pre>

<p>Both <code class="language-plaintext highlighter-rouge">SyntaxError</code> and <code class="language-plaintext highlighter-rouge">IndentationError</code> indicate a problem with the syntax of your program,
but an <code class="language-plaintext highlighter-rouge">IndentationError</code> is more specific:
it <em>always</em> means that there is a problem with how your code is indented.</p>

<p>:::::::::::::::::::::::::::::::::::::::::  callout</p>

<h2 id="tabs-and-spaces">Tabs and Spaces</h2>

<p>Some indentation errors are harder to spot than others.
In particular, mixing spaces and tabs can be difficult to spot
because they are both <a href="../learners/reference.md#whitespace">whitespace</a>.
In the example below, the first two lines in the body of the function
<code class="language-plaintext highlighter-rouge">some_function</code> are indented with tabs, while the third line — with spaces.
If you’re working in a Jupyter notebook, be sure to copy and paste this example
rather than trying to type it in manually because Jupyter automatically replaces
tabs with spaces.</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="k">def</span> <span class="nf">some_function</span><span class="p">():</span>
	<span class="n">msg</span> <span class="o">=</span> <span class="s">'hello, world!'</span>
	<span class="k">print</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">msg</span>
</code></pre></div></div>

<p>Visually it is impossible to spot the error.
Fortunately, Python does not allow you to mix tabs and spaces.</p>

<pre><code class="language-error">  File "&lt;ipython-input-5-653b36fbcd41&gt;", line 4
    return msg
              ^
TabError: inconsistent use of tabs and spaces in indentation
</code></pre>

<p>::::::::::::::::::::::::::::::::::::::::::::::::::</p>

<h2 id="variable-name-errors">Variable Name Errors</h2>

<p>Another very common type of error is called a <code class="language-plaintext highlighter-rouge">NameError</code>,
and occurs when you try to use a variable that does not exist.
For example:</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="k">print</span><span class="p">(</span><span class="n">a</span><span class="p">)</span>
</code></pre></div></div>

<pre><code class="language-error">---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
&lt;ipython-input-7-9d7b17ad5387&gt; in &lt;module&gt;()
----&gt; 1 print(a)

NameError: name 'a' is not defined
</code></pre>

<p>Variable name errors come with some of the most informative error messages,
which are usually of the form “name ‘the_variable_name’ is not defined”.</p>

<p>Why does this error message occur?
That’s a harder question to answer,
because it depends on what your code is supposed to do.
However,
there are a few very common reasons why you might have an undefined variable.
The first is that you meant to use a
<a href="../learners/reference.md#string">string</a>, but forgot to put quotes around it:</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="k">print</span><span class="p">(</span><span class="n">hello</span><span class="p">)</span>
</code></pre></div></div>

<pre><code class="language-error">---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
&lt;ipython-input-8-9553ee03b645&gt; in &lt;module&gt;()
----&gt; 1 print(hello)

NameError: name 'hello' is not defined
</code></pre>

<p>The second reason is that you might be trying to use a variable that does not yet exist.
In the following example,
<code class="language-plaintext highlighter-rouge">count</code> should have been defined (e.g., with <code class="language-plaintext highlighter-rouge">count = 0</code>) before the for loop:</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="k">for</span> <span class="n">number</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">10</span><span class="p">):</span>
    <span class="n">count</span> <span class="o">=</span> <span class="n">count</span> <span class="o">+</span> <span class="n">number</span>
<span class="k">print</span><span class="p">(</span><span class="s">'The count is:'</span><span class="p">,</span> <span class="n">count</span><span class="p">)</span>
</code></pre></div></div>

<pre><code class="language-error">---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
&lt;ipython-input-9-dd6a12d7ca5c&gt; in &lt;module&gt;()
      1 for number in range(10):
----&gt; 2     count = count + number
      3 print('The count is:', count)

NameError: name 'count' is not defined
</code></pre>

<p>Finally, the third possibility is that you made a typo when you were writing your code.
Let’s say we fixed the error above by adding the line <code class="language-plaintext highlighter-rouge">Count = 0</code> before the for loop.
Frustratingly, this actually does not fix the error.
Remember that variables are <a href="../learners/reference.md#case-sensitive">case-sensitive</a>,
so the variable <code class="language-plaintext highlighter-rouge">count</code> is different from <code class="language-plaintext highlighter-rouge">Count</code>. We still get the same error,
because we still have not defined <code class="language-plaintext highlighter-rouge">count</code>:</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">Count</span> <span class="o">=</span> <span class="mi">0</span>
<span class="k">for</span> <span class="n">number</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">10</span><span class="p">):</span>
    <span class="n">count</span> <span class="o">=</span> <span class="n">count</span> <span class="o">+</span> <span class="n">number</span>
<span class="k">print</span><span class="p">(</span><span class="s">'The count is:'</span><span class="p">,</span> <span class="n">count</span><span class="p">)</span>
</code></pre></div></div>

<pre><code class="language-error">---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
&lt;ipython-input-10-d77d40059aea&gt; in &lt;module&gt;()
      1 Count = 0
      2 for number in range(10):
----&gt; 3     count = count + number
      4 print('The count is:', count)

NameError: name 'count' is not defined
</code></pre>

<h2 id="index-errors">Index Errors</h2>

<p>Next up are errors having to do with containers (like lists and strings) and the items within them.
If you try to access an item in a list or a string that does not exist,
then you will get an error.
This makes sense:
if you asked someone what day they would like to get coffee,
and they answered “caturday”,
you might be a bit annoyed.
Python gets similarly annoyed if you try to ask it for an item that doesn’t exist:</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">letters</span> <span class="o">=</span> <span class="p">[</span><span class="s">'a'</span><span class="p">,</span> <span class="s">'b'</span><span class="p">,</span> <span class="s">'c'</span><span class="p">]</span>
<span class="k">print</span><span class="p">(</span><span class="s">'Letter #1 is'</span><span class="p">,</span> <span class="n">letters</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
<span class="k">print</span><span class="p">(</span><span class="s">'Letter #2 is'</span><span class="p">,</span> <span class="n">letters</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
<span class="k">print</span><span class="p">(</span><span class="s">'Letter #3 is'</span><span class="p">,</span> <span class="n">letters</span><span class="p">[</span><span class="mi">2</span><span class="p">])</span>
<span class="k">print</span><span class="p">(</span><span class="s">'Letter #4 is'</span><span class="p">,</span> <span class="n">letters</span><span class="p">[</span><span class="mi">3</span><span class="p">])</span>
</code></pre></div></div>

<pre><code class="language-output">Letter #1 is a
Letter #2 is b
Letter #3 is c
</code></pre>

<pre><code class="language-error">---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
&lt;ipython-input-11-d817f55b7d6c&gt; in &lt;module&gt;()
      3 print('Letter #2 is', letters[1])
      4 print('Letter #3 is', letters[2])
----&gt; 5 print('Letter #4 is', letters[3])

IndexError: list index out of range
</code></pre>

<p>Here,
Python is telling us that there is an <code class="language-plaintext highlighter-rouge">IndexError</code> in our code,
meaning we tried to access a list index that did not exist.</p>

<h2 id="file-errors">File Errors</h2>

<p>The last type of error we’ll cover today
are those associated with reading and writing files: <code class="language-plaintext highlighter-rouge">FileNotFoundError</code>.
If you try to read a file that does not exist,
you will receive a <code class="language-plaintext highlighter-rouge">FileNotFoundError</code> telling you so.
If you attempt to write to a file that was opened read-only, Python 3
returns an <code class="language-plaintext highlighter-rouge">UnsupportedOperationError</code>.
More generally, problems with input and output manifest as
<code class="language-plaintext highlighter-rouge">OSError</code>s, which may show up as a more specific subclass; you can see
<a href="https://docs.python.org/3/library/exceptions.html#os-exceptions">the list in the Python docs</a>.
They all have a unique UNIX <code class="language-plaintext highlighter-rouge">errno</code>, which is you can see in the error message.</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">file_handle</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="s">'myfile.txt'</span><span class="p">,</span> <span class="s">'r'</span><span class="p">)</span>
</code></pre></div></div>

<pre><code class="language-error">---------------------------------------------------------------------------
FileNotFoundError                         Traceback (most recent call last)
&lt;ipython-input-14-f6e1ac4aee96&gt; in &lt;module&gt;()
----&gt; 1 file_handle = open('myfile.txt', 'r')

FileNotFoundError: [Errno 2] No such file or directory: 'myfile.txt'
</code></pre>

<p>One reason for receiving this error is that you specified an incorrect path to the file.
For example,
if I am currently in a folder called <code class="language-plaintext highlighter-rouge">myproject</code>,
and I have a file in <code class="language-plaintext highlighter-rouge">myproject/writing/myfile.txt</code>,
but I try to open <code class="language-plaintext highlighter-rouge">myfile.txt</code>,
this will fail.
The correct path would be <code class="language-plaintext highlighter-rouge">writing/myfile.txt</code>.
It is also possible that the file name or its path contains a typo.</p>

<p>A related issue can occur if you use the “read” flag instead of the “write” flag.
Python will not give you an error if you try to open a file for writing
when the file does not exist.
However,
if you meant to open a file for reading,
but accidentally opened it for writing,
and then try to read from it,
you will get an <code class="language-plaintext highlighter-rouge">UnsupportedOperation</code> error
telling you that the file was not opened for reading:</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">file_handle</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="s">'myfile.txt'</span><span class="p">,</span> <span class="s">'w'</span><span class="p">)</span>
<span class="n">file_handle</span><span class="p">.</span><span class="n">read</span><span class="p">()</span>
</code></pre></div></div>

<pre><code class="language-error">---------------------------------------------------------------------------
UnsupportedOperation                      Traceback (most recent call last)
&lt;ipython-input-15-b846479bc61f&gt; in &lt;module&gt;()
      1 file_handle = open('myfile.txt', 'w')
----&gt; 2 file_handle.read()

UnsupportedOperation: not readable
</code></pre>

<p>These are the most common errors with files,
though many others exist.
If you get an error that you’ve never seen before,
searching the Internet for that error type
often reveals common reasons why you might get that error.</p>

<p>:::::::::::::::::::::::::::::::::::::::  challenge</p>

<h2 id="identifying-syntax-errors">Identifying Syntax Errors</h2>

<ol>
  <li>Read the code below, and (without running it) try to identify what the errors are.</li>
  <li>Run the code, and read the error message. Is it a <code class="language-plaintext highlighter-rouge">SyntaxError</code> or an <code class="language-plaintext highlighter-rouge">IndentationError</code>?</li>
  <li>Fix the error.</li>
  <li>Repeat steps 2 and 3, until you have fixed all the errors.</li>
</ol>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="k">def</span> <span class="nf">another_function</span>
  <span class="k">print</span><span class="p">(</span><span class="s">'Syntax errors are annoying.'</span><span class="p">)</span>
   <span class="k">print</span><span class="p">(</span><span class="s">'But at least Python tells us about them!'</span><span class="p">)</span>
  <span class="k">print</span><span class="p">(</span><span class="s">'So they are usually not too hard to fix.'</span><span class="p">)</span>
</code></pre></div></div>

<p>:::::::::::::::  solution</p>

<h2 id="solution-1">Solution</h2>

<p><code class="language-plaintext highlighter-rouge">SyntaxError</code> for missing <code class="language-plaintext highlighter-rouge">():</code> at end of first line,
<code class="language-plaintext highlighter-rouge">IndentationError</code> for mismatch between second and third lines.
A fixed version is:</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="k">def</span> <span class="nf">another_function</span><span class="p">():</span>
    <span class="k">print</span><span class="p">(</span><span class="s">'Syntax errors are annoying.'</span><span class="p">)</span>
    <span class="k">print</span><span class="p">(</span><span class="s">'But at least Python tells us about them!'</span><span class="p">)</span>
    <span class="k">print</span><span class="p">(</span><span class="s">'So they are usually not too hard to fix.'</span><span class="p">)</span>
</code></pre></div></div>

<p>:::::::::::::::::::::::::</p>

<p>::::::::::::::::::::::::::::::::::::::::::::::::::</p>

<p>:::::::::::::::::::::::::::::::::::::::  challenge</p>

<h2 id="identifying-variable-name-errors">Identifying Variable Name Errors</h2>

<ol>
  <li>Read the code below, and (without running it) try to identify what the errors are.</li>
  <li>Run the code, and read the error message.
  What type of <code class="language-plaintext highlighter-rouge">NameError</code> do you think this is?
  In other words, is it a string with no quotes,
  a misspelled variable,
  or a variable that should have been defined but was not?</li>
  <li>Fix the error.</li>
  <li>Repeat steps 2 and 3, until you have fixed all the errors.</li>
</ol>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="k">for</span> <span class="n">number</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">10</span><span class="p">):</span>
    <span class="c1"># use a if the number is a multiple of 3, otherwise use b
</span>    <span class="k">if</span> <span class="p">(</span><span class="n">Number</span> <span class="o">%</span> <span class="mi">3</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
        <span class="n">message</span> <span class="o">=</span> <span class="n">message</span> <span class="o">+</span> <span class="n">a</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">message</span> <span class="o">=</span> <span class="n">message</span> <span class="o">+</span> <span class="s">'b'</span>
<span class="k">print</span><span class="p">(</span><span class="n">message</span><span class="p">)</span>
</code></pre></div></div>

<p>:::::::::::::::  solution</p>

<h2 id="solution-2">Solution</h2>

<p>3 <code class="language-plaintext highlighter-rouge">NameError</code>s for <code class="language-plaintext highlighter-rouge">number</code> being misspelled, for <code class="language-plaintext highlighter-rouge">message</code> not defined,
and for <code class="language-plaintext highlighter-rouge">a</code> not being in quotes.</p>

<p>Fixed version:</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">message</span> <span class="o">=</span> <span class="s">''</span>
<span class="k">for</span> <span class="n">number</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">10</span><span class="p">):</span>
    <span class="c1"># use a if the number is a multiple of 3, otherwise use b
</span>    <span class="k">if</span> <span class="p">(</span><span class="n">number</span> <span class="o">%</span> <span class="mi">3</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
        <span class="n">message</span> <span class="o">=</span> <span class="n">message</span> <span class="o">+</span> <span class="s">'a'</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">message</span> <span class="o">=</span> <span class="n">message</span> <span class="o">+</span> <span class="s">'b'</span>
<span class="k">print</span><span class="p">(</span><span class="n">message</span><span class="p">)</span>
</code></pre></div></div>

<p>:::::::::::::::::::::::::</p>

<p>::::::::::::::::::::::::::::::::::::::::::::::::::</p>

<p>:::::::::::::::::::::::::::::::::::::::  challenge</p>

<h2 id="identifying-index-errors">Identifying Index Errors</h2>

<ol>
  <li>Read the code below, and (without running it) try to identify what the errors are.</li>
  <li>Run the code, and read the error message. What type of error is it?</li>
  <li>Fix the error.</li>
</ol>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">seasons</span> <span class="o">=</span> <span class="p">[</span><span class="s">'Spring'</span><span class="p">,</span> <span class="s">'Summer'</span><span class="p">,</span> <span class="s">'Fall'</span><span class="p">,</span> <span class="s">'Winter'</span><span class="p">]</span>
<span class="k">print</span><span class="p">(</span><span class="s">'My favorite season is '</span><span class="p">,</span> <span class="n">seasons</span><span class="p">[</span><span class="mi">4</span><span class="p">])</span>
</code></pre></div></div>

<p>:::::::::::::::  solution</p>

<h2 id="solution-3">Solution</h2>

<p><code class="language-plaintext highlighter-rouge">IndexError</code>; the last entry is <code class="language-plaintext highlighter-rouge">seasons[3]</code>, so <code class="language-plaintext highlighter-rouge">seasons[4]</code> doesn’t make sense.
A fixed version is:</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">seasons</span> <span class="o">=</span> <span class="p">[</span><span class="s">'Spring'</span><span class="p">,</span> <span class="s">'Summer'</span><span class="p">,</span> <span class="s">'Fall'</span><span class="p">,</span> <span class="s">'Winter'</span><span class="p">]</span>
<span class="k">print</span><span class="p">(</span><span class="s">'My favorite season is '</span><span class="p">,</span> <span class="n">seasons</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
</code></pre></div></div>

<p>:::::::::::::::::::::::::</p>

<p>::::::::::::::::::::::::::::::::::::::::::::::::::</p>

<p>:::::::::::::::::::::::::::::::::::::::: keypoints</p>

<ul>
  <li>Tracebacks can look intimidating, but they give us a lot of useful information about what went wrong in our program, including where the error occurred and what type of error it was.</li>
  <li>An error having to do with the ‘grammar’ or syntax of the program is called a <code class="language-plaintext highlighter-rouge">SyntaxError</code>. If the issue has to do with how the code is indented, then it will be called an <code class="language-plaintext highlighter-rouge">IndentationError</code>.</li>
  <li>A <code class="language-plaintext highlighter-rouge">NameError</code> will occur when trying to use a variable that does not exist. Possible causes are that a variable definition is missing, a variable reference differs from its definition in spelling or capitalization, or the code contains a string that is missing quotes around it.</li>
  <li>Containers like lists and strings will generate errors if you try to access items in them that do not exist. This type of error is called an <code class="language-plaintext highlighter-rouge">IndexError</code>.</li>
  <li>Trying to read a file that does not exist will give you an <code class="language-plaintext highlighter-rouge">FileNotFoundError</code>. Trying to read a file that is open for writing, or writing to a file that is open for reading, will give you an <code class="language-plaintext highlighter-rouge">IOError</code>.</li>
</ul>

<p>::::::::::::::::::::::::::::::::::::::::::::::::::</p>

